File: dotnet/src/Pactman.NonprofitCheckPlus/PactmanEnvironment.cs
using System;
using System.Collections.Generic;

namespace Pactman.NonprofitCheckPlus
{
    /// <summary>
    /// Named Pactman environments that are supported for public SDK use.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Endpoint hosts are declared here and nowhere else. Nothing else in this
    /// package contains a literal Pactman host.
    /// </para>
    /// <para>
    /// Pactman's QA, SIT and sandbox hosts are internal and are deliberately not
    /// exposed. Point at one with <see cref="Configuration.PactmanClientOptions.BaseUrl"/> if you
    /// have been given access to it.
    /// </para>
    /// </remarks>
    public enum PactmanEnvironment
    {
        /// <summary>The Pactman production API.</summary>
        Production = 0,
    }

    /// <summary>Base URLs and name parsing for <see cref="PactmanEnvironment"/>.</summary>
    public static class PactmanEnvironments
    {
        /// <summary>The environment used when none is supplied.</summary>
        public const PactmanEnvironment Default = PactmanEnvironment.Production;

        /// <summary>The base URL every request in an environment is sent to.</summary>
        /// <param name="environment">The environment to resolve.</param>
        /// <returns>The absolute base URL, with no trailing slash.</returns>
        /// <exception cref="ArgumentOutOfRangeException">The value is not a declared environment.</exception>
        public static string BaseUrl(this PactmanEnvironment environment) => environment switch
        {
            PactmanEnvironment.Production => "https://entities.pactman.org",
            _ => throw new ArgumentOutOfRangeException(nameof(environment), environment, "Unknown Pactman environment."),
        };

        /// <summary>The lower-case wire name of an environment, as the other Pactman SDKs spell it.</summary>
        /// <param name="environment">The environment to name.</param>
        /// <returns>The environment's canonical name.</returns>
        /// <exception cref="ArgumentOutOfRangeException">The value is not a declared environment.</exception>
        public static string Name(this PactmanEnvironment environment) => environment switch
        {
            PactmanEnvironment.Production => "production",
            _ => throw new ArgumentOutOfRangeException(nameof(environment), environment, "Unknown Pactman environment."),
        };

        /// <summary>Every environment name the SDK understands.</summary>
        /// <returns>The supported environments, in declaration order.</returns>
        public static IReadOnlyList<PactmanEnvironment> Supported() => new[] { PactmanEnvironment.Production };

        /// <summary>The supported names, joined for an error message.</summary>
        /// <returns>A comma-separated list of environment names.</returns>
        public static string SupportedNames() => string.Join(", ", Array.ConvertAll(new[] { PactmanEnvironment.Production }, Name));

        /// <summary>
        /// Parses an environment name, case-insensitively.
        /// </summary>
        /// <param name="name">The name to parse, such as <c>production</c>.</param>
        /// <param name="environment">The parsed environment when this returns <see langword="true"/>.</param>
        /// <returns><see langword="true"/> when the name is a supported environment.</returns>
        public static bool TryParse(string? name, out PactmanEnvironment environment)
        {
            environment = Default;

            if (string.IsNullOrWhiteSpace(name))
            {
                return false;
            }

            if (string.Equals(name!.Trim(), "production", StringComparison.OrdinalIgnoreCase))
            {
                environment = PactmanEnvironment.Production;
                return true;
            }

            return false;
        }
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/PactmanEnvironment.java
package org.pactman.nonprofitcheckplus;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

/**
 * A named Pactman environment.
 *
 * <p>Endpoint hosts are declared here and nowhere else. Nothing in this package
 * should contain a literal Pactman host.
 *
 * <p>Pactman's QA, SIT and sandbox hosts are internal and are deliberately not
 * exposed. Point at one with
 * {@link org.pactman.nonprofitcheckplus.config.PactmanClientOptions.Builder#baseUrl(String)}
 * if you have been given access to it.
 */
public enum PactmanEnvironment {

    /** The Pactman production API. */
    PRODUCTION("production", "https://entities.pactman.org");

    private final String environmentName;
    private final String baseUrl;

    PactmanEnvironment(String environmentName, String baseUrl) {
        this.environmentName = environmentName;
        this.baseUrl = baseUrl;
    }

    /**
     * The environment's name as the SDKs in every language spell it.
     *
     * @return the lowercase name, such as {@code production}.
     */
    public String environmentName() {
        return environmentName;
    }

    /**
     * The base URL requests to this environment are sent to.
     *
     * @return an absolute origin, with no trailing slash.
     */
    public String baseUrl() {
        return baseUrl;
    }

    /**
     * Looks an environment up by name, case-insensitively.
     *
     * @param name the environment name, such as {@code production}.
     * @return the environment, or {@code null} when no environment has that name.
     */
    public static PactmanEnvironment fromName(String name) {
        if (name == null) {
            return null;
        }

        String needle = name.trim().toLowerCase(java.util.Locale.ROOT);

        for (PactmanEnvironment environment : values()) {
            if (environment.environmentName.equals(needle)) {
                return environment;
            }
        }

        return null;
    }

    /**
     * Every environment name the SDK understands.
     *
     * @return the supported names, in declaration order.
     */
    public static List<String> supportedNames() {
        return Collections.unmodifiableList(
                java.util.Arrays.stream(values())
                        .map(PactmanEnvironment::environmentName)
                        .collect(Collectors.toList()));
    }

    @Override
    public String toString() {
        return environmentName;
    }
}
